torch/fx/passes/dialect/common/cse_pass.py

```diff
@@ -0,0 +1,90 @@
+from typing import Dict, Tuple, Any
+
+import torch
+import torch.fx as fx
+from torch.fx.passes.infra.pass_base import PassBase, PassResult
+from torch.utils._pytree import tree_flatten
+
+aten = torch.ops.aten
+
+
+def get_aten_target(node):
+    if hasattr(node.target, 'overloadpacket'):
+        return node.target.overloadpacket
+    return node.target
+
+# random ops are banned from CSE
+rand_ops = [aten.dropout, aten._fused_dropout, aten._standard_gamma,
+            aten.bernoulli, aten.multinomial, aten.native_dropout,
+            aten.normal, aten.poisson, aten.binomial, aten.rrelu,
+            aten.rand_like, aten.rand, aten.randint, aten.randn, aten.randperm]
+
+
+class CSEPass(PassBase):
+
+    def call(self, graph_module: fx.GraphModule) -> PassResult:
+        """
+        Return a new copy of torch.fx.GraphModule with CSE applied to the input graph
+
+        Example usage:
+
+        from functorch import make_fx
+        def f(a):
+            b = a * a
+            c = a * a
+            return b+c
+
+        p = CSEPass()
+        traced_graph = make_fx(f)(torch.tensor(1))
+        print(traced_graph)
+        result = p(traced_graph)
+        print(result.graph_module)
+        """
+        modified = False
+        new_graph = fx.Graph()
+        env: Dict[fx.node.Node, fx.node.Node] = {}  # map from node in the old graph to node in the new graph
+        hash_env: Dict[Tuple[torch._ops.OpOverload, int], fx.node.Node] = {}  # map from hash to a node in the new graph
+        token_map: Dict[Tuple[torch._ops.OpOverload, int], Dict[str, Any]] = {}  # map from hash to token
+        for n in graph_module.graph.nodes:
+            # The placeholder, output, and get_attr nodes are copied to the new grpah without change
+            # do not CSE away random operations
+            if n.op == 'placeholder' or n.op == 'output' or n.op == 'get_attr' or get_aten_target(n) in rand_ops:
+                new_node = new_graph.node_copy(n, lambda x: env[x])
+                env[n] = new_node
+            else:  # n.op == 'call_function', should never see n.op == 'call_module' or 'call_method'
+                # substitute args and kwargs memebrs to their mapping in env if exists
+                # specs can be used to reconstruct nested list/dictionaries
+                def substitute(arg_list):
+                    arg_list, spec = tree_flatten(arg_list)
+                    for i in range(len(arg_list)):
+                        v = arg_list[i]
+                        if isinstance(v, torch.fx.node.Node) and v in env:
+                            arg_list[i] = env[v]
+                    return tuple(arg_list), spec
+                args, args_spec = substitute(n.args)
+                kwargs, kwargs_spec = substitute(n.kwargs)
+
+                # each token corresponds to a unique node
```

Comment:
Sorry... having a hard time following the logic here... 

Does it handle following case?

```
def fn(x):
     # branch 1
     a = x + 1
     b = a + 2
     # branch 2
    c = x + 1
    d = c + 3
    return b, d 

two branches have the identical subgraph, and should be CSEed. 








```

Reply:
yeah it will be handled.  `c` will be replaced with `a`

Reply:
oh do you mean d = c+2? It will also be handled.

First `c` will be replaced with `a` when we process `c`. On line 80, env[node c in old graph] = node a in new graph.

Then `d` will be replaced with `a+2` when we process `d` on line 64-65 because `c` is mapped to `a` in `env`, now we see that it is identical with `b=a+2` on line 73-78.

Reply:
Try the following:

```
def f(x):
    # branch 1
    a = x + 1
    b = a + 2
     # branch 2
    c = x + 1
    d = c + 2
    return b, d 

p = CSEPass()
traced_graph = make_fx(f, decomposition_table=default_decompositions)(torch.tensor(1))
print(traced_graph)
result = p(traced_graph)
print(result.graph_module)
print(result.modified)
```

the output is

f()



def forward(self, x_1):
    add_tensor = torch.ops.aten.add.Tensor(x_1, 1)
    add_tensor_1 = torch.ops.aten.add.Tensor(add_tensor, 2);  add_tensor = None
    add_tensor_2 = torch.ops.aten.add.Tensor(x_1, 1);  x_1 = None
    add_tensor_3 = torch.ops.aten.add.Tensor(add_tensor_2, 2);  add_tensor_2 = None
    return (add_tensor_1, add_tensor_3)
    
GraphModule()



def forward(self, x_1):
    add_tensor = torch.ops.aten.add.Tensor(x_1, 1);  x_1 = None
    add_tensor_1 = torch.ops.aten.add.Tensor(add_tensor, 2);  add_tensor = None
    return (add_tensor_1, add_tensor_1)
    
True

Reply:
Note that if you do `c = x` and `d=c+3` it cannot detect that `d==b`.  All the args of nodes have to be identical. It doesn't do math...

Reply:
I meant +2